### PR TITLE
Fix needs_quotes() in `to nuon` (closes #6989)

### LINKED ISSUE #6989: `['4'] | to nuon` produces `[4]` (turns it into a list<int>)
### Describe the bug

See title.

### How to reproduce

```
〉['4'] | to nuon
[4]
〉['4'] | describe
list<string>
〉['4'] | to nuon | from nuon | describe
list<int>
```
See the first example and third example.

### Expected behavior

Same behaviour as this (but without whitespace):
```
〉['4'] | to json
[
  "4"
]
```

### Screenshots

_No response_

### Configuration

| key                | value                                           |
| ------------------ | ----------------------------------------------- |
| version            | 0.70.0                                          |
| branch             |                                                 |
| commit_hash        | 9ef65dcd692b502d7476b1787247fae8638c2f0c        |
| build_os           | windows-x86_64                                  |
| build_target       | x86_64-pc-windows-msvc                          |
| rust_version       | rustc 1.64.0 (a55dd71d5 2022-09-19)             |
| rust_channel       | stable-x86_64-pc-windows-msvc                   |
| cargo_version      | cargo 1.64.0 (387270bc7 2022-09-16)             |
| pkg_version        | 0.70.0                                          |
| build_time         | 2022-10-18 18:55:02 +00:00                      |
| build_rust_channel | release                                         |
| features           | database, dataframe, default, trash, which, zip |
| installed_plugins  |                                                 |

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by webbedspace**
# Description

Closes #6989. Adds an intensive test to ensure strings that could match various non-string literals are always serialised with quotes.

Also:

* null now serialises as `null` instead of `$nothing` (required because Nuon is a JSON superset)
* Improved wording of 'not supported' error for block-to-nuon and others.

# Major Changes

If you're considering making any major change to nushell, before starting work on it, seek feedback from regular contributors and get approval for the idea from the core team either on [Discord](https://discordapp.com/invite/NtAbbGn) or [GitHub issue](https://github.com/nushell/nushell/issues/new/choose).
Making sure we're all on board with the change saves everybody's time.
Thanks!

# Tests + Formatting

Make sure you've done the following, if applicable:

- Add tests that cover your changes (either in the command examples, the crate/tests folder, or in the /tests folder)
  - Try to think about corner cases and various ways how your changes could break. Cover those in the tests

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- `cargo test --workspace --features=extra` to check that all tests pass

# After Submitting

* Help us keep the docs up to date: If your PR affects the user experience of Nushell (adding/removing a command, changing an input/output type, etc.), make sure the changes are reflected in the documentation (https://github.com/nushell/nushell.github.io) after the PR is merged.

### CODE REVIEW

File: crates/nu-command/src/formats/to/nuon.rs [hunk 1]
```diff
@@ -189,26 +196,22 @@ fn to_nuon(call: &Call, input: PipelineData) -> Result<String, ShellError> {
...
+    // • Any character of []:`{}#';()|$,
+    //   • The above hits Datetimes via hitting :
+    // • Any whitespace
+    // • true, false, null
+    // • inf, Infinity, nan
+    // • Integers and floats
+    // • Filesizes
+    // • Durations
+    // • Ranges
+    static ref NEED_QUOTE_REGEX: Regex = Regex::new(r#"(?i)[\[\]:`\{\}#'";\(\)\|\$,\s]|^(true|false|null|[+-]?inf(inity)?|[+-]?nan)$|^[+-]?\d+\.?\d*(([kmgtpez]?i?b)|[mnu]s|sec|min|hr|day|wk)?$|^([+-]?\d+\.?\d*)?\.\.<?([+-]?\d+\.?\d*)?$"#).expect("internal error: NEED_QUOTE_REGEX didn't compile");
```
Comment (sholderbach) [inline crates/nu-command/src/formats/to/nuon.rs:210]: Wow that's a mouthful of a regex :D
Do we need a `fancy_regex` to do this? If we need backreferences, in the worst case you might be able to craft DoS strings.
The other question would be if we would be willing to accept a simpler heuristic with some false positives if this turns out to be a bottleneck in emitting nuon.

Comment (webbedspace) [reply crates/nu-command/src/formats/to/nuon.rs:210]: This is already relatively simple (it false-positives on various invalid ranges like `..<` and invalid capitalisations like `TRUE`, so that it can hit every capitalisation of `nan`, `infinity`, filesize suffixes etc.).

This has no backreferences so I could presumably turn it into a "vanilla" regex.

Comment (sholderbach) [reply crates/nu-command/src/formats/to/nuon.rs]: If I remember correctly `fancy_regex` builds on top of `regex` so pulling this in might not be an additional dep.

But optimization of the performance can be a second step, if we start to benchmark this codepath.

Comment (dandavison) [reply crates/nu-command/src/formats/to/nuon.rs]: I think @sholderbach is right and I think that we should prefer `regex` over `fancy_regex` everywhere possible, and I think it would be nice -- if this regex doesn't use `fancy_regex` features -- to express it using `regex` because that would prove that it doesn't have any of the undesirable runtime complexities that `fancy_regex` allows.

Comment (webbedspace) [reply crates/nu-command/src/formats/to/nuon.rs]: Well… `fancy-regex` does "build upon" `regex`, but as a fork of `regex` rather than something that uses it as an explicit dep. It does [promise](https://docs.rs/fancy-regex/latest/fancy_regex/) to have the same performance of `regex` when none of [this syntax](https://docs.rs/fancy-regex/latest/fancy_regex/#syntax) is in the regex… and, like, you can just read the regex and see none of that stuff appears in it. To me, explicitly bringing in the `regex` crate doesn't seem like it would accomplish much.

File: crates/nu-command/src/formats/to/nuon.rs [hunk 2]
```diff
@@ -1,4 +1,6 @@
 use core::fmt::Write;
+use fancy_regex::Regex;
+use lazy_static::lazy_static;
```
Comment (rgwood) [inline crates/nu-command/src/formats/to/nuon.rs:3]: Going forward, we generally want to use `once_cell` instead of `lazy_static`. `once_cell` is probably going to be in the stdlib eventually and I believe the ecosystem has started to standardize on it accordingly: https://github.com/rust-lang/rust/issues/74465

File: crates/nu-command/src/formats/to/nuon.rs (hunk 1 above)
Comment (fdncred) [reply crates/nu-command/src/formats/to/nuon.rs]: To be clear, when I was researching `fancy-regex` if not "fancy" features are used, then it uses regex by default.

**Comment by rgwood** [comment]
Thanks for the PR!

I'm personally not a big fan of doing this all in 1 huge regex. It means I can no longer follow the code without using an external tool like [regexper](https://regexper.com).

I wonder if it would make sense to refactor that 1 regex into multiple smaller functions/regexes.

**Comment by sholderbach** [comment]
OK what I gather from the discussion: regexes are hard, (good to know that fancy_regex will use the simpler implementation if possible)

But parsing that regex manually is non trivial (I got regex dyslexic at some point, why I was asking about backreferences).
This has negative implications for maintainability (or if someone were to try to write their own `nuon` (de)serializer.)

Furthermore it can also not fully capture the complexity around nushell literals as you said.
I think avoiding the quotes is foremost an optimization to avoid sending too many quotes over the wire but not critical to the function of the format to serialize and deserialize faithfully.

From that standpoint (and the possible implementation) perspective I am kind of inclined to say that we should go for simpler `str.contains` or trivial matches to make sure all the roundtrip test pass (i.e. no false negatives), but accept some false positives.

Regexes are certainly performant for common prefixes and many possible paths through the automaton but heurisitcally there are probably some features that are more likely to disqualify bare strings that wouldn't need the full regex (e.g. space).

**Comment by webbedspace** [comment]
I have now reduced the regex down to just a single character-class check. Most notably, this now considers any digit to be grounds for adding quotes (thus auto-hitting every number-based special type, filesizes, ranges, datetimes, durations).

Also I swapped in `once_cell` as advised.

**Comment by fdncred** [comment]
I just tested this and it seems to work as advertised. I still hate regexes but at least it's smaller.

**Comment by rgwood** [comment]
Yeah, looks good to go IMO. @sholderbach are you OK if we merge this?

**Comment by webbedspace** [comment]
Um, is this ready to be merged?

File: crates/nu-command/src/formats/to/nuon.rs [hunk 3]
```diff
@@ -189,26 +196,29 @@ fn to_nuon(call: &Call, input: PipelineData) -> Result<String, ShellError> {
...
+    // These are case-sensitive keywords
+    match string {
+        "true" | "false" | "null" => return true,
+        _ => (),
+    };
+    // These are case-insensitive keywords
+    match string.to_lowercase().as_str() {
+        "inf" | "infinity" | "nan" | "+inf" | "+infinity" | "+nan" | "-inf" | "-infinity"
+        | "-nan" => return true,
+        _ => (),
```
Comment (sholderbach) [inline crates/nu-command/src/formats/to/nuon.rs:218]: This allocates unnecessarily (especially for long strings that would be rejected due to spaces). 
Either a regex due to shared substrings or simpler https://doc.rust-lang.org/stable/std/primitive.str.html#method.eq_ignore_ascii_case

**Review by sholderbach — COMMENTED** [review]
Ah sorry, totally forgot to get back on this one. Was wondering too much if we could build a realistic corpus for benchmarks but that shouldn't be the scope of this PR. 

Small nit that can certainly be addressed without exact benchmarks.

File: crates/nu-command/src/formats/to/nuon.rs (hunk 3 above)
Comment (webbedspace) [reply crates/nu-command/src/formats/to/nuon.rs]: OK, I've moved the inf/infinity/nan cases back into NEEDS_QUOTES_REGEX.

**Comment by sholderbach** [comment]
Thanks! That should really make the nuon format more robust!
